client/core/core.go

```diff
@@ -1922,6 +1911,68 @@ func (c *Core) WalletState(assetID uint32) *WalletState {
 	return wallet.state()
 }
 
+func (c *Core) walletCheckAndNotify(w *xcWallet) bool {
+	synced, progress, err := w.SyncStatus()
+	if err != nil {
+		c.log.Errorf("Unable to get wallet/node sync status for %s: %v",
+			unbip(w.AssetID), err)
+		return false
+	}
+	w.mtx.Lock()
+	w.synced = synced
+	w.syncProgress = progress
+	w.mtx.Unlock()
+	c.notify(newWalletStateNote(w.state()))
+	if synced {
+		c.updateWalletBalance(w)
+		c.log.Infof("Wallet synced for asset %s", unbip(w.AssetID))
+	}
+	return synced
+}
+
+// RescanWallet will issue a rescan command to the wallet if supported by the
+// wallet implementation. It is up to the underlying wallet backend if and how
+// to implement this functionality. It may be asynchronous. Core will emit
+// wallet state notifications until the rescan is complete. WARNING: It is
+// ill-advised to initiate a wallet rescan with active orders unless as a last
+// ditch effort to get the wallet to recognize a transaction needed to complete
+// a swap.
+func (c *Core) RescanWallet(assetID uint32) error {
+	wallet, err := c.connectedWallet(assetID)
+	if err != nil {
+		return fmt.Errorf("OpenWallet: wallet not found for %d -> %s: %w",
+			assetID, unbip(assetID), err)
+	}
+
+	// Begin potentially asynchronous wallet rescan operation.
+	if err = wallet.Rescan(c.ctx); err != nil {
+		return err
+	}
+
+	if c.walletCheckAndNotify(wallet) {
+		return nil // sync done
+	}
```

Comment:
Would we ever expect this to be the case?

Reply:
Depends if the `wallet.Rescan` implementation is async or not... or at least that was my thinking.  I wasn't ready to formalize `Rescan` as necessarily (a)synchronous.  Hmmm